modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/model.scala

```diff
@@ -4,388 +4,275 @@ import scalaz._
     *
-    * - (Execution, Option[Sequence]) \/ x:
+    * If after marking the `Action`, all elements in `Current` are `Result`, it
+    * empties `Current` and moves the `Result` to the completed `Queue`
     *
-    * Returns the next `Execution` and the remaining `Sequence` when there are
-    * no more `Execution`s in the current `Step`. If this was the last `Step` of
-    * the current `Sequence` it becomes `None`.
+    * If the index doesn't exist, the new `State` is returned unmodified.
     */
-  // TODO: Handle unrooted executions
-  def uncons[A](seq: Sequence[A]): (((Execution[A], Int), Option[Sequence[A]]) \/ (Execution[A], Sequence[A])) = {
-    val (exe, mstep) = Step.uncons(seq.steps.head)
-    mstep match {
-      // No more Executions in Step, remove Step
-      case None => (
-        (exe, seq.steps.head.id),
-        seq.steps.tail.toNel.map(Sequence(seq.id, _, seq.executions))
-      ).left
-      // More Executions in Step, remove Execution in Step
-      case Some(st) => (exe, head.set(seq, st)).right
-    }
-  }
+  def mark(i: Int)(r: Result): QState
+
 }
 
-/**
-  * A list of `Executions` grouped by observation.
-  */
-case class Step[A](id: Int, executions: NonEmptyList[Execution[A]])
+case class QStateI(pending: Queue[Action], status: Status) extends QState { self =>
 
-object Step {
+  val next: Option[QState] =
+    Zipper.QueueZ.currentify(pending).map(QStateZ(_, status))
 
-  def executions[A]: Step[A] @> NonEmptyList[Execution[A]] =
-    Lens.lensu((s, exes) => s.copy(executions = exes), _.executions)
+  val current: Current = Current.empty
 
-  def head[A]: Step[A] @> Execution[A] = executions >=> Lens.nelHeadLens
+  val done: Queue[Result] = Queue(Nil)
 
-  /**
-    * Adds a `Execution` to the front of a `Step`.
-    */
-  def cons[A](exe: Execution[A])(st: Step[A]): Step[A] = Step(st.id, exe <:: st.executions)
+  def mark(i: Int)(r: Result): QState = self
 
-  /**
-    * Return the next `Execution` and the remaining `Step` if there are more
-    * `Execution`s left.
-    */
-  def uncons[A](st: Step[A]): (Execution[A], Option[Step[A]]) =
-    (st.executions.head,
-     st.executions.tail.toNel.map(Step(st.id, _))
-    )
 }
 
-/**
-  * This structure holds the `Execution` currently under execution. It carries
-  * information about which `Action`s have been completed and the position in
-  * the `Queue` for proper insertion into the completed `Queue` when all the
-  * `Execution`s are done.
-  */
-case class Current(
-  ars:  Vector[Action \/ Result],
-  ctxt: Option[(String, Int) \/ Int]
-) {
+case class QStateZ(zipper: Zipper.QueueZ, status: Status) extends QState { self =>
 
-  def isEmpty: Boolean = ars.empty
+  val next: Option[QState] = zipper.next match {
+    // Last execution
+    case None => zipper.uncurrentify.map(QStateF(_, status))
+    case Some(x) => Some(QStateZ(x, status))
+  }
 
-  def actions: List[Action] = {
+  val current: Current = zipper.focus.focus.focus
```

Comment:
`current` is the current Execution. I'll put a comment about which each focus level refers to. Outside of this module you should never need to mess up with `focus`, only `current`.